File: braccio_gamepad_teleop/src/gamepad_teleop.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/joy.hpp>
#include <trajectory_msgs/msg/joint_trajectory.hpp>
#include <trajectory_msgs/msg/joint_trajectory_point.hpp>
#include <builtin_interfaces/msg/duration.hpp>
#include <array>
#include <algorithm>

class GamepadTeleopSimple : public rclcpp::Node
{
public:
    GamepadTeleopSimple() : Node("gamepad_teleop_simple")
    {
        // Posiciones iniciales basadas en el estado real del robot en Gazebo
        current_joint_positions_ = {1.573, 0.763, 3.142, 3.127, 1.572, 0.0};

        RCLCPP_INFO(this->get_logger(), "🎮 Gamepad Teleop Simple iniciado");
        RCLCPP_INFO(this->get_logger(), "📐 Posiciones iniciales (estado real): [%.3f, %.3f, %.3f, %.3f, %.3f]", 
                   current_joint_positions_[0], current_joint_positions_[1], current_joint_positions_[2], 
                   current_joint_positions_[3], current_joint_positions_[4]);

        // Subscriber para mensajes de joystick
        joy_subscription_ = this->create_subscription<sensor_msgs::msg::Joy>(
            "/joy", 10, std::bind(&GamepadTeleopSimple::joy_callback, this, std::placeholders::_1));

        // Publisher directo usando el mismo método que pick_and_place
        trajectory_publisher_ = this->create_publisher<trajectory_msgs::msg::JointTrajectory>(
            "/position_trajectory_controller/joint_trajectory", 10);



        RCLCPP_INFO(this->get_logger(), "✅ Publisher configurado usando método pick_and_place");
        RCLCPP_INFO(this->get_logger(), "🎮 Usa el stick izquierdo para controlar base y joint_1");
        RCLCPP_INFO(this->get_logger(), "🎮 Usa el stick derecho para controlar joint_2 y joint_3"); 
        RCLCPP_INFO(this->get_logger(), "🎮 Usa L1/R1 para controlar joint_4");
        RCLCPP_INFO(this->get_logger(), "📍 Modo: Incremental desde estado real del robot");
    }

private:
    void joy_callback(const sensor_msgs::msg::Joy::SharedPtr msg)
    {
        // Deadzone para evitar drift
        const double deadzone = 0.1;

        // Mapeo de controles (PlayStation/Xbox layout)
        double left_x = std::abs(msg->axes[0]) > deadzone ? msg->axes[0] : 0.0;  // joint_base
        double left_y = std::abs(msg->axes[1]) > deadzone ? msg->axes[1] : 0.0;  // joint_1
        double right_x = std::abs(msg->axes[3]) > deadzone ? msg->axes[3] : 0.0; // joint_2
        double right_y = std::abs(msg->axes[4]) > deadzone ? msg->axes[4] : 0.0; // joint_3

        // Botones para joint_4
        double joint_4_input = 0.0;
        if (msg->buttons.size() > 5) {
            if (msg->buttons[4]) joint_4_input = 1.0;  // L1 -> positivo
            if (msg->buttons[5]) joint_4_input = -1.0; // R1 -> negativo
        }
        // Botones para gripper (right_gripper_joint)
        double gripper_input = 0.0;
        if (msg->buttons.size() > 7) {
            if (msg->buttons[6]) gripper_input = 1.0;  // L2 -> abrir
            if (msg->buttons[7]) gripper_input = -1.0; // R2 -> cerrar
        }
        // Solo procesar si hay algún input
        if (std::abs(left_x) > 0 || std::abs(left_y) > 0 || 
            std::abs(right_x) > 0 || std::abs(right_y) > 0 || 
            std::abs(joint_4_input) > 0 || std::abs(gripper_input) > 0) {
            // Factor de velocidad ajustado para suavidad y rapidez
            const double velocity_factor = 0.005; // Incremento menor para suavidad
            // Aplicar incrementos
            current_joint_positions_[0] += left_x * velocity_factor;  // joint_base
            current_joint_positions_[1] += left_y * velocity_factor;  // joint_1
            current_joint_positions_[2] += right_y * velocity_factor; // joint_2 (stick Y)
            current_joint_positions_[3] += right_x * velocity_factor; // joint_3 (stick X)
            current_joint_positions_[4] += joint_4_input * velocity_factor; // joint_4
            current_joint_positions_[5] += gripper_input * velocity_factor; // right_gripper_joint
            // Aplicar límites de seguridad
            apply_joint_limits();
            // Enviar comandos separados para brazo y gripper
            send_arm_trajectory_command();
            send_gripper_trajectory_command();
            // Log de debug con valores actualizados (menos frecuente para suavidad visual)
            RCLCPP_INFO_THROTTLE(this->get_logger(), *this->get_clock(), 1000, 
                "🎮 Input: LX=%.2f LY=%.2f RX=%.2f RY=%.2f J4=%.2f G=%.2f → Pos: [%.3f, %.3f, %.3f, %.3f, %.3f, %.3f]",
                left_x, left_y, right_x, right_y, joint_4_input, gripper_input,
                current_joint_positions_[0], current_joint_positions_[1], current_joint_positions_[2], 
                current_joint_positions_[3], current_joint_positions_[4], current_joint_positions_[5]);
        }
    }
    
    void apply_joint_limits()
    {
        // Límites de seguridad para cada joint (en radianes)
        const std::array<double, 6> min_limits = {0, 0.4, 0.0, 0.0, 0.0, 0.1};
        const std::array<double, 6> max_limits = {3.14, 2.7, 3.14, 3.14, 3.14, 1.0};
        for (size_t i = 0; i < 6; ++i) {
            current_joint_positions_[i] = std::clamp(
                current_joint_positions_[i], min_limits[i], max_limits[i]);
        }
    }
    
    void send_arm_trajectory_command()
    {
        // Trajectory for arm (first 5 joints)
        auto traj = trajectory_msgs::msg::JointTrajectory();
        traj.joint_names = {
            "joint_base",
            "joint_1",
            "joint_2",
            "joint_3",
            "joint_4"
        };
        auto point = trajectory_msgs::msg::JointTrajectoryPoint();
        point.positions = {
            current_joint_positions_[0],
            current_joint_positions_[1],
            current_joint_positions_[2],
            current_joint_positions_[3],
            current_joint_positions_[4]
        };
        point.time_from_start = builtin_interfaces::msg::Duration();
        point.time_from_start.sec = 0;
        point.time_from_start.nanosec = 50000000;
        traj.points.push_back(point);
        trajectory_publisher_->publish(traj);
        RCLCPP_DEBUG(this->get_logger(), "📤 Trayectoria brazo enviada");
    }

    void send_gripper_trajectory_command()
    {
        // Trajectory for gripper (right_gripper_joint)
        auto traj = trajectory_msgs::msg::JointTrajectory();
        traj.joint_names = { "right_gripper_joint" };
        auto point = trajectory_msgs::msg::JointTrajectoryPoint();
        point.positions = { current_joint_positions_[5] };
        point.time_from_start = builtin_interfaces::msg::Duration();
        point.time_from_start.sec = 0;
        point.time_from_start.nanosec = 20000000;
        traj.points.push_back(point);
        if (!gripper_trajectory_publisher_) {
            gripper_trajectory_publisher_ = this->create_publisher<trajectory_msgs::msg::JointTrajectory>(
                "/gripper_controller/joint_trajectory", 10);
        }
        gripper_trajectory_publisher_->publish(traj);
        RCLCPP_DEBUG(this->get_logger(), "📤 Trayectoria gripper enviada");
    }

    // Variables miembro
    rclcpp::Subscription<sensor_msgs::msg::Joy>::SharedPtr joy_subscription_;
    rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr trajectory_publisher_;
    rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr gripper_trajectory_publisher_;
    std::array<double, 6> current_joint_positions_;
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<GamepadTeleopSimple>();
    
    RCLCPP_INFO(node->get_logger(), "🚀 Iniciando teleoperación simple con gamepad...");
    
    rclcpp::spin(node);
    rclcpp::shutdown();
    
    return 0;
}

File: braccio_gamepad_teleop/scripts/gamepad_teleop.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Joy
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
from builtin_interfaces.msg import Duration
from linkattacher_msgs.srv import AttachLink, DetachLink
from gazebo_msgs.msg import ModelStates
import tf_transformations
from tf2_ros import Buffer, TransformListener
import math

class GamepadTeleopSimple(Node):
    def __init__(self):
        super().__init__('gamepad_teleop_simple')
        self.current_joint_positions = [1.573, 0.763, 3.142, 3.127, 1.572, 0.0]
        self.get_logger().info('🎮 Gamepad Teleop Simple iniciado')
        self.get_logger().info(f'📐 Posiciones iniciales (estado real): {self.current_joint_positions}')
        self.joy_subscription = self.create_subscription(
            Joy, '/joy', self.joy_callback, 10)
        self.arm_pub = self.create_publisher(
            JointTrajectory, '/position_trajectory_controller/joint_trajectory', 10)
        self.gripper_pub = self.create_publisher(
            JointTrajectory, '/gripper_controller/joint_trajectory', 10)
        self.attach_client = self.create_client(AttachLink, '/ATTACHLINK')
        self.detach_client = self.create_client(DetachLink, '/DETACHLINK')
        self.get_logger().info('✅ Publisher configurado usando método pick_and_place')
        self.get_logger().info('🎮 Usa el stick izquierdo para controlar base y joint_1')
        self.get_logger().info('🎮 Usa el stick derecho para controlar joint_2 y joint_3')
        self.get_logger().info('🎮 Usa L1/R1 para controlar joint_4')
        self.get_logger().info('⭕ Usa Círculo para pick and place (attach)')
        self.get_logger().info('❌ Usa X para detach (soltar cubo)')
        self.get_logger().info('📍 Modo: Incremental desde estado real del robot')

        # Para almacenar posiciones de modelos
        self.model_states = None
        self.model_states_sub = self.create_subscription(
            ModelStates, '/gazebo/model_states', self.model_states_callback, 10)

        # Nombre del link del gripper (frame TF)
        self.gripper_link_name = 'right_gripper_link'
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

    def model_states_callback(self, msg):
        self.model_states = msg

    def joy_callback(self, msg):
        deadzone = 0.1
        left_x = msg.axes[0] if abs(msg.axes[0]) > deadzone else 0.0
        left_y = msg.axes[1] if abs(msg.axes[1]) > deadzone else 0.0
        right_x = msg.axes[3] if abs(msg.axes[3]) > deadzone else 0.0
        right_y = msg.axes[4] if abs(msg.axes[4]) > deadzone else 0.0
        joint_4_input = 0.0
        if len(msg.buttons) > 5:
            if msg.buttons[4]: joint_4_input = 1.0
            if msg.buttons[5]: joint_4_input = -1.0
        gripper_input = 0.0
        if len(msg.buttons) > 7:
            if msg.buttons[6]: gripper_input = 1.0
            if msg.buttons[7]: gripper_input = -1.0
        # Pick basado en proximidad (sin requerir contacto físico)
        if len(msg.buttons) > 2 and msg.buttons[2]:
            model, link = self.get_closest_object_to_gripper()
            if model:
                # Verificar que el objeto esté lo suficientemente cerca (umbral de proximidad)
                try:
                    trans = self.tf_buffer.lookup_transform('world', self.gripper_link_name, rclpy.time.Time())
                    gripper_pos = (trans.transform.translation.x, trans.transform.translation.y, trans.transform.translation.z)
                    
                    # Buscar la posición del objeto más cercano
                    for i, name in enumerate(self.model_states.name):
                        if name == model:
                            obj_pose = self.model_states.pose[i]
                            obj_pos = (obj_pose.position.x, obj_pose.position.y, obj_pose.position.z)
                            dist = math.sqrt(sum((a-b)**2 for a, b in zip(gripper_pos, obj_pos)))
                            
                            # Umbral de proximidad para pick (15cm)
                            if dist <= 0.15:
                                self.get_logger().info(f'🎯 Pick por proximidad: {model} (distancia: {dist:.3f}m)')
                                self.call_attach_service(model, link)
                            else:
                                self.get_logger().warn(f'❌ Objeto {model} demasiado lejos para pick: {dist:.3f}m (máx: 0.15m)')
                            break
                except Exception as e:
                    self.get_logger().warn(f'Error calculando proximidad para pick: {e}')
            else:
                self.get_logger().warn('No se encontró objeto cercano para pick')
        if len(msg.buttons) > 0 and msg.buttons[0]:
            model, link = self.get_closest_object_to_gripper()
            if model:
                self.get_logger().info(f'Modelo más cercano para detach: {model}, link: {link}')
                self.call_detach_service(model, link)
            else:
                self.get_logger().warn('No se encontró objeto cercano para detach')

        # Movimiento del brazo y gripper siempre funciona
        if any(abs(x) > 0 for x in [left_x, left_y, right_x, right_y, joint_4_input, gripper_input]):
            velocity_factor = 0.01
            self.current_joint_positions[0] += left_x * velocity_factor
            self.current_joint_positions[1] += left_y * velocity_factor
            self.current_joint_positions[2] += right_y * velocity_factor
            self.current_joint_positions[3] += right_x * velocity_factor
            self.current_joint_positions[4] += joint_4_input * velocity_factor
            self.current_joint_positions[5] += gripper_input * velocity_factor
            self.apply_joint_limits()
            self.send_arm_trajectory_command()
            self.send_gripper_trajectory_command()
            self.get_logger().info(
                f'🎮 Input: LX={left_x:.2f} LY={left_y:.2f} RX={right_x:.2f} RY={right_y:.2f} J4={joint_4_input:.2f} G={gripper_input:.2f} → Pos: {self.current_joint_positions}')

    def get_closest_object_to_gripper(self):
        # Espera a tener datos de modelos SOLO para pick/detach
        if self.model_states is None:
            self.get_logger().warn('model_states aún no disponible')
            return None, None
        self.get_logger().info(f'Modelos en model_states: {self.model_states.name}')
        # Obtén la posición del gripper usando TF
        try:
            trans = self.tf_buffer.lookup_transform('world', self.gripper_link_name, rclpy.time.Time())
            gripper_pos = (
                trans.transform.translation.x,
                trans.transform.translation.y,
                trans.transform.translation.z
            )
            self.get_logger().info(f'Posición gripper TF: {gripper_pos}')
        except Exception as e:
            self.get_logger().warn(f'No se pudo obtener la posición del gripper por TF: {e}')
            return None, None
        min_dist = float('inf')
        closest_model = None
        for i, name in enumerate(self.model_states.name):
            if name in ['braccio', 'ground_plane']:
                continue
            obj_pose = self.model_states.pose[i]
            obj_pos = (obj_pose.position.x, obj_pose.position.y, obj_pose.position.z)
            self.get_logger().info(f'Comparando con modelo: {name}, posición: {obj_pos}')
            dist = math.sqrt(sum((a-b)**2 for a, b in zip(gripper_pos, obj_pos)))
            self.get_logger().info(f'Distancia a {name}: {dist}')
            if dist < min_dist:
                min_dist = dist
                closest_model = name
        self.get_logger().info(f'Modelo más cercano: {closest_model}, distancia: {min_dist}')
        return closest_model, 'link'

    def call_attach_service(self, model2_name, link2_name):
        if not self.attach_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().warn('Servicio /ATTACHLINK no disponible')
            return
        req = AttachLink.Request()
        req.model1_name = 'braccio'
        req.link1_name = 'right_gripper_link'
        req.model2_name = model2_name
        req.link2_name = link2_name
        future = self.attach_client.call_async(req)
        self.get_logger().info(f'📎 Intentando attach: {model2_name}/{link2_name}')

    def call_detach_service(self, model2_name, link2_name):
        if not self.detach_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().warn('Servicio /DETACHLINK no disponible')
            return
        req = DetachLink.Request()
        req.model1_name = 'braccio'
        req.link1_name = 'right_gripper_link'
        req.model2_name = model2_name
        req.link2_name = link2_name
        future = self.detach_client.call_async(req)
        self.get_logger().info(f'🔗 Intentando detach: {model2_name}/{link2_name}')

    def apply_joint_limits(self):
        min_limits = [0, 0.4, 0.0, 0.0, 0.0, 0.1]
        max_limits = [3.14, 2.7, 3.14, 3.14, 3.14, 1.0]
        self.current_joint_positions = [max(min(val, max_limits[i]), min_limits[i]) for i, val in enumerate(self.current_joint_positions)]

    def send_arm_trajectory_command(self):
        traj = JointTrajectory()
        traj.joint_names = ['joint_base', 'joint_1', 'joint_2', 'joint_3', 'joint_4']
        point = JointTrajectoryPoint()
        point.positions = self.current_joint_positions[:5]
        point.time_from_start = Duration(sec=0, nanosec=50000000)
        traj.points.append(point)
        self.arm_pub.publish(traj)

    def send_gripper_trajectory_command(self):
        traj = JointTrajectory()
        traj.joint_names = ['right_gripper_joint']
        point = JointTrajectoryPoint()
        point.positions = [self.current_joint_positions[5]]
        point.time_from_start = Duration(sec=0, nanosec=20000000)
        traj.points.append(point)
        self.gripper_pub.publish(traj)

def main(args=None):
    rclpy.init(args=args)
    node = GamepadTeleopSimple()
    node.get_logger().info('🚀 Iniciando teleoperación simple con gamepad...')
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
